lib/resources/oracledb_session.rb

```diff
@@ -13,30 +16,110 @@ class OracledbSession < Inspec.resource(1)
+
+      # we prefer sqlci although it is way slower than sqlplus, but it understands csv properly
+      @sqlcl_bin = 'sql'
       @sqlplus_bin = opts[:sqlplus_bin] || 'sqlplus'
-      return skip_resource("Can't run Oracle checks without authentication") if @user.nil? or @pass.nil?
+
+      return skip_resource "Can't run Oracle checks without authentication" if @user.nil? or @password.nil?
+      return skip_resource 'You must provide at least an SID and/or a Service Name for the session' if @service.nil?
     end
 
     def query(q)
       escaped_query = q.gsub(/\\/, '\\\\').gsub(/"/, '\\"')
-      cmd = inspec.command("echo \"#{escaped_query}\" | #{@sqlplus_bin} -s #{@user}/#{@pass}@#{@host}/#{@sid}")
+      # escape tables with $
+      escaped_query = escaped_query.gsub('$', "\\$")
+
+      p = nil
+      # check if sqlcl is available and prefer that
+      if inspec.command(@sqlcl_bin).exist?
+        bin = @sqlcl_bin
+        opts = "set sqlformat csv\nSET FEEDBACK OFF"
+        p = 'parse_csv_result'
+      elsif inspec.command(@sqlplus_bin).exist?
+        bin = @sqlplus_bin
+        opts = "SET MARKUP HTML ON\nSET FEEDBACK OFF"
+        p = 'parse_html_result'
+      end
+
+      return skip_resource("Can't find suitable Oracle CLI") if p.nil?
+
+      cmd = inspec.command("echo \"#{opts}\n#{escaped_query};\nEXIT\" | #{bin} -s #{@user}/#{@password}@#{@host}:#{@port}/#{@service}")
       out = cmd.stdout + "\n" + cmd.stderr
       if out.downcase =~ /^error/
-        skip_resource("Can't connect to Oracle instance for SQL checks.")
+        return skip_resource("Can't connect to Oracle instance for SQL checks.")
       end
 
-      cmd
+      SQLQueryResult.new(cmd, send(p.to_sym, cmd.stdout.gsub(/\r/,'')))
     end
 
     def to_s
       'Oracle Session'
     end
+
+    private
+
+    def parse_csv_result(stdout)
+      require 'csv'
+      table = CSV.parse(stdout, { headers: true })
+
+      # convert to hash
+      headers = table.headers
+
+      results = table.map { |row|
+        res = {}
+        headers.each { |header|
+          res[header.downcase] = row[header]
+        }
+        Hashie::Mash.new(res)
+      }
+      results
+    end
+
+    def parse_html_result(stdout)
+      result = stdout
+      # make oracle html valid html by removing the p tag, it does not include a closing tag
+      result = result.gsub('<p>', '').gsub('</p>', '').gsub('<br>', '')
+
+      require 'rexml/document'
+      doc = REXML::Document.new result
+      table = doc.elements['table']
+
+      hash = {}
+      if !table.nil?
+        rows = table.elements.to_a
+        headers = rows[0].elements.to_a('/th').map { |entry| entry.text.strip }
+        rows.delete(0)
+
+        # iterate over each row, first row is header
+        hash = rows.map { |row|
+          res = {}
+          entries = row.elements.to_a('/td')
+
+          headers.each_with_index { |header, index|
+            # TODO: once nokogiri is back, we can remove html entities
```

Comment:
Are there plans to bring `nokogiri` back in? It has only caused us pain on multiple projects, especially since we support Windows. Can we just remove this TODO and continue to use `htmlentities` if we need it?

I don't suppose there's a non-HTML markup we can use with Oracle to avoid this?